### ISSUE TITLE
System.FormatException: The response was not valid.

### ISSUE THREAD

**Issue by EvanMulawski** [EvanMulawski/FanControl.CorsairLink#5]
> I just got home from work and tried the latest build, unfortunately I'm still getting errors. Here is my log.txt. Maybe I'm running a different firmware version.

```
2023-01-16 7:09:18 PM: [CorsairLink] DeviceManager: Found 7 Corsair device(s)
  name=, devicePath=\\?\hid#virtualdevice&10&col01#2&3b8f46cf&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=, devicePath=\\?\hid#virtualdevice&10&col05#2&3b8f46cf&0&0004#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=, devicePath=\\?\hid#virtualdevice&10&col02#2&3b8f46cf&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd
  name=, devicePath=\\?\hid#virtualdevice&10&col03#2&3b8f46cf&0&0002#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd
  name=, devicePath=\\?\hid#virtualdevice&10&col04#2&3b8f46cf&0&0003#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_01#7&387a1dd5&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

2023-01-16 7:09:18 PM: [CorsairLink] DeviceManager: Found 1 supported Corsair device(s)
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

2023-01-16 7:09:18 PM: [CorsairLink] An exception occurred attempting to initialize device 'CORSAIR iCUE Commander CORE (630f0001801ccabae51f2af52091005f)' (\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}):
2023-01-16 7:09:18 PM: [CorsairLink] System.FormatException: The response was not valid.
   at CorsairLink.CommanderCoreDevice.EndpointResponse.ThrowIfInvalid()
   at CorsairLink.CommanderCoreDevice.InitializeSpeedChannels(EndpointResponse hardwareFixedSpeedPercentResponse)
   at CorsairLink.CommanderCoreDevice.Initialize()
   at CorsairLink.CommanderCoreDevice.Connect()
   at FanControl.CorsairLink.CorsairLinkPlugin.FanControl.Plugins.IPlugin.Initialize()
2023-01-16 7:09:20 PM: [CorsairLink] DeviceManager: Found 7 Corsair device(s)
  name=, devicePath=\\?\hid#virtualdevice&10&col01#2&3b8f46cf&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=, devicePath=\\?\hid#virtualdevice&10&col05#2&3b8f46cf&0&0004#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=, devicePath=\\?\hid#virtualdevice&10&col02#2&3b8f46cf&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd
  name=, devicePath=\\?\hid#virtualdevice&10&col03#2&3b8f46cf&0&0002#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd
  name=, devicePath=\\?\hid#virtualdevice&10&col04#2&3b8f46cf&0&0003#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_01#7&387a1dd5&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

2023-01-16 7:09:20 PM: [CorsairLink] DeviceManager: Found 1 supported Corsair device(s)
  name=CORSAIR iCUE Commander CORE, devicePath=\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

2023-01-16 7:09:20 PM: [CorsairLink] An exception occurred attempting to initialize device 'CORSAIR iCUE Commander CORE (630f0001801ccabae51f2af52091005f)' (\\?\hid#vid_1b1c&pid_0c1c&mi_00#7&20b69190&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}):
2023-01-16 7:09:20 PM: [CorsairLink] System.FormatException: The response was not valid.
   at CorsairLink.CommanderCoreDevice.EndpointResponse.ThrowIfInvalid()
   at CorsairLink.CommanderCoreDevice.InitializeSpeedChannels(EndpointResponse hardwareFixedSpeedPercentResponse)
   at CorsairLink.CommanderCoreDevice.Initialize()
   at CorsairLink.CommanderCoreDevice.Connect()
   at FanControl.CorsairLink.CorsairLinkPlugin.FanControl.Plugins.IPlugin.Initialize()
2023-01-16 7:09:22 PM: CorsairLink could not initialize or has no sensors.
```

_Originally posted by @Lithradus in https://github.com/EvanMulawski/FanControl.CorsairLink/issues/2#issuecomment-1384688620_

**Comment by EvanMulawski** [comment]
@Lithradus Do you have any other software running (e.g. SignalRGB, OpenRGB, liquidctl, etc.) that could be attempting to communicate with the device? This error is usually due to another program reading from and/or writing to the device.

**Comment by Lithradus** [comment]
The only software I can think of that might interfere is iCUE which is disabled with the services set to manual and LHM which isn't set to start with windows and wasn't started before testing. In case it makes any difference I'm running Windows 11 with the latest dev channel flight. That alone could cause any number of issues.

**Comment by EvanMulawski** [comment]
Please make sure the device is running firmware version >= `2.10.219`. You'll need to launch iCUE to check and update.

**Comment by Lithradus** [comment]
It should be on the latest but I'll check it after work

**Comment by Lithradus** [comment]
The firmware version is 2.11.221

**Comment by dshk0718** [comment]
@EvanMulawski I did some digging around this, and for some reason `hardwareFixedSpeedPercentResponse` does not get the right data type and thus does not match the `DataTypes.HardwareSpeedFixedPercent` type when checking for sequence equality. That ultimately leads to this `System.FormatException: The response was not valid.` Otherwise, all the other readings returned from the other responses are all correctly read (`connectedSpeedsResponse`, `speedsResponse`, and `temperaturesResponse`).

What is this `hardwareFixedSpeedPercentResponse` and what is it used for?

**Comment by dshk0718** [comment]
Another update... @EvanMulawski I am not so sure how this happens, but it seems that the AIO cooler loses its firmware? When I opened up iCUE and updated the firmware (since it prompted me and said it's mandatory), it started working again.

**Comment by EvanMulawski** [comment]
@dshk0718 The `HardwareSpeedFixedPercent` endpoint returns the requested percent speed/power currently stored in the device. I'm using it during initialization to keep the fan speeds where they were before Fan Control starts instead of setting them to a default value like I have to on the PRO. I'm not able to reproduce the error on my CORE XT, so this might only be a problem on the CORE non-XT.

There might be two firmwares depending on your AIO model, one just called "Firmware" (which is the CORE) and the other is "LCD Screen Firmware." Can you post a screenshot of what you're seeing in iCUE?

**Comment by dshk0718** [comment]
![image](https://user-images.githubusercontent.com/22249997/213355615-e0d631de-b084-47c3-a2a2-4b5281bcf68b.png)
@EvanMulawski  Is this what you are looking for? Btw, I do not have the LCD screen on my AIO.

**Comment by EvanMulawski** [comment]
@dshk0718 Can you configure a hardware lighting profile in iCUE and save it to the device? That should allow you to have some RGB customization while running Fan Control.

**Comment by dshk0718** [comment]
> @dshk0718 Can you configure a hardware lighting profile in iCUE and save it to the device? That should allow you to have some RGB customization while running Fan Control.
@EvanMulawski Yup, that's what I did. To note though, like @Lithradus mentioned in the other issue thread, it does blink every few seconds. I assume that's because your plugin is accessing the controller to get all the appropriate fan speeds and whatnot and causing an I/O interrupt on the controller (which is in turn causing this blinking issue)?

**Comment by EvanMulawski** [comment]
Reopened #4 - will investigate.

**Comment by EvanMulawski** [comment]
@dshk0718 Have you seen this error again since updating the firmware?

**Comment by dshk0718** [comment]
No, I have not so far, as long as iCUE (and its related services) and SignalRGB stay closed and not running.

**Comment by EvanMulawski** [comment]
This should be fixed in **v1.0.0-beta.6**.

Note: If this does occur, it will now appear as `CorsairLink.CorsairLinkDeviceException: Operation canceled. The expected data type was not read within the specified time.` in the log.
